### ISSUE TITLE
./configure says nanosleep not required, then quits because of it

### ISSUE THREAD

**Issue by ShaunCurrier** [esnet/iperf#503]
While running `./configure --host=i686-w64-mingw32`, the following is printed to console and it quits early:
```
<snip>
checking for library containing nanosleep... none required
<snip>
nanosleep() required for timing operations.
```

So, it says a nanosleep library is not required. Then says it is required.  Then it quits.

**Comment by bmah888** [comment]
mingw is not an explicitly supported platform.  That having been said I'm not sure why it's behaving the way it is.  What version of iperf3 are you trying to build?

**Comment by scurrier03** [comment]
I have `iperf-3.1.4`, which I got from the following link.
http://downloads.es.net/pub/iperf/iperf-3-current.tar.gz

**Comment by scurrier03** [comment]
Related thread in the mingw-w64 discussion forum.
https://sourceforge.net/p/mingw-w64/discussion/723798/thread/7f8ca21d/?limit=25

**Comment by scurrier03** [comment]
I've never dealt with shell scripts like `configure` before, but I noticed something that seems odd.  There are two places in the script where the error message `nanosleep() required for timing operations` appears.

The first occurrence, it's in a block of code that seems to be heavily related to `nanosleep` because it's surrounded by variables named after `nanosleep` and the command at the top of the block is `# Solaris puts nanosleep in -lrt`.  This seems totally normal.

The second occurrence, it's in a block of code that is similar to the first, but has replaced `nanosleep` with `hstrerror` - all except for the line where a reference to `nanosleep` remains: `echo "nanosleep() required for timing operations."`.

In my novice view, it looks like some code was copied and all references to `nanosleep` replaced, but the error message mentioning `nanosleep` was not replaced.

I see the two occurrences of the text `nanosleep() required for timing operations` on line 11827 and 11890 of the `configure` file.  This file doesn't appear in the repo.  I'm guessing it's generated before being offered at the link I referred to above?

I'm not sure how exactly it relates, but perhaps this ultimately traces back to [line 74](https://github.com/esnet/iperf/blob/01fb3e6dcfe91d5870c9b2187d1deff7aca95341/configure.ac#L74) in [configure.ac](https://github.com/esnet/iperf/blob/01fb3e6dcfe91d5870c9b2187d1deff7aca95341/configure.ac).

So maybe this is all just a case of a copy/paste error leading to an incorrect error message?

**Comment by Borim7** [comment]
I have stumbled over the same problem when building iperf with MSYS2. Shaun as you discovered the "nanosleep required" message is used twice and the second occurrence is the problematic one. So the actual problem is hstrerror.

When you check the man page for hstrerror this function is marked as deprecated. Also from a quick investigation: This function is never used.

```
if (h_errno && herr) {
        strncat(errstr, hstrerror(h_errno), len - strlen(errstr) - 1);
} else*/ if (errno && perr) {
        strncat(errstr, strerror(errno), len - strlen(errstr) - 1);
}
```

and h_errno is always 0. At least with a text search I found only the line `h_errno = 0;` and no other occurrence.

How about removing hstrerror from iperf? If that is ok I can do it and the issue with the socket function. Thats the next issue you will hit after the solving the hstrerror problem.

**Comment by bmah888** [comment]
I needed to dig into the usage of h_errno and hstrerror(3) to understand why the "dead code" assertion, it wasn't obvious at all to me why.  The issue is not that we never set h_errno in our code.  Really, h_errno is set as a side-effect of a gethostbyname(3) call that failed.  That said gethostbyname(3) is deprecated, and if we ever had any calls to that function in iperf3, they seem to have all been replaced by getaddrinfo(3), which is actually the preferred alternative.  The equivalent to hstrerror(h_errno) is roughly to call gai_strerror(3) with the return code from getaddrinfo(3), but we don't do that.  If we did, we might be able to provide some better name resolution diagnostics.  But that's a different enhancement.

Bottom line:  I agree that hstrerror(3) can never be called, but it's for slightly different reasons than was indicated a few comments ago.

Also relevant to #523.

**Comment by bmah888** [comment]
Pull request #523, which fixes this issue, merged to the mainline.  Should be merged to the 3.1-STABLE release branch.  I will keep this issue open to track that.

**Comment by ShaunCurrier** [comment]
I think the socket issue that @borim7 mentioned is still going to be present, but maybe we can open a new issue for that for organization's sake.

**Comment by bmah888** [comment]
Merged.  Closing.
